### PR TITLE
AP_Scripting: add Plane autoland applet

### PR DESCRIPTION
**Pull request by Hwurzburg**
Facilitates having an autoland FS mission setup for flying from fields without prior planning(or computers).....sets  up an autolanding sequence based on takeoff direction....

https://youtu.be/Y8Hn5h5h-3s?si=d7UuL5IL1rAcW1ZM

### CODE REVIEW

File: libraries/AP_Scripting/applets/UniversalAutoLand.lua [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    return update,1000
+  end
+  
+  if not mission_loaded then --if first time after arm and enabled is valid then create  mission 
+    local home = ahrs:get_home()
+    local location = ahrs:get_location()
+    if location:alt() > (home:alt() + param:get("TKOFF_LVL_ALT")*100) then
+        local yaw = ahrs:get_yaw()
+        mission:set_item(2,create_do_land_start_WP())
+        mission:set_item(3,create_final_approach_WP(3,math.deg(yaw)+180,FIN_WP_DIST,param:get("TKOFF_ALT")))
```
Comment (timtuxworth) [inline libraries/AP_Scripting/applets/UniversalAutoLand.lua:102]: You might want to wrap the yaw after adding 180 (e.g. 359 degrees + 180 = 539), Tridges scripted aerobatics has some wrap functions that will do it.

**Comment by Hwurzburg** [comment]
yeah, I thought about that but forgot....will wrap it

    On Monday, November 11, 2024 at 10:17:39 PM CST, Tim Tuxworth ***@***.***> wrote:  
 
 


@timtuxworth commented on this pull request.

In libraries/AP_Scripting/applets/UniversalAutoLand.lua:
> +   return item
+end
+
+function update()
+  if not arming:is_armed() then --if disarmed, wait until armed
+    mission_loaded = false
+    return update,1000
+  end
+  
+  if not mission_loaded then --if first time after arm and enabled is valid then create  mission 
+    local home = ahrs:get_home()
+    local location = ahrs:get_location()
+    if location:alt() > (home:alt() + param:get("TKOFF_LVL_ALT")*100) then
+        local yaw = ahrs:get_yaw()
+        mission:set_item(2,create_do_land_start_WP())
+        mission:set_item(3,create_final_approach_WP(3,math.deg(yaw)+180,FIN_WP_DIST,param:get("TKOFF_ALT")))

You might want to wrap the yaw after adding 180 (e.g. 359 degrees + 180 = 539), Tridges scripted aerobatics has some wrap functions that will do it.

File: libraries/AP_Scripting/applets/UniversalAutoLand.lua (hunk 1 above)
Comment (Hwurzburg) [reply libraries/AP_Scripting/applets/UniversalAutoLand.lua]: had planned on it but forgot while debugging...done

**Comment by Hwurzburg** [comment]
@timtuxworth yeah I lost the mission reset somewhere along the way...made a few tweaks also...felt I needed the user to consciously set the SCR_USER params instead of providing a default if they did not...

**Comment by Rolf-G** [comment]
I have read the instructions 10 times, but have not yet understood how exactly the landing direction is determined ? (I don't mean in the LUA script, but actually in flight operations)

**Comment by Hwurzburg** [comment]
Ahrs has a get yaw function that returns vehicle heading

File: libraries/AP_Scripting/applets/UniversalAutoLand.lua [hunk 2]
```diff
@@ -0,0 +1,139 @@
+--[[ Upon Arming , creates a four item mission consisting of: NAV_TAKEOFF, DO_LAND_START,Final Approach WP opposite bearing from HOME of heading used during takeoff, at TKOFF_ALT or SCR_USER3 above home, SCR_USER2 or 2X TKOFF_DIST, and a LAND waypoint at HOME and stops until next disarm/boot. SCR_USER1 is used to enable or disable it.
+--]]
+
+--local variablees
+
+SCR_USER_EN = Parameter("SCR_USER1") -- enable this script
```
Comment (tridge) [inline libraries/AP_Scripting/applets/UniversalAutoLand.lua:6]: should create a param table for an applet I think, with param docs

File: libraries/AP_Scripting/applets/UniversalAutoLand.lua [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        local yaw = ahrs:get_yaw()
+        mission:set_item(3,create_final_approach_WP(3,wrap_180(math.deg(yaw)+180),FIN_WP_DIST,FIN_WP_ALT))
+        mission:set_item(4,create_land_WP())
+              mission_loaded = true
+    end
+  end
+  return update, 1000
+end 
+
+gcs:send_text(5,"Loaded TKOFF/RTL/Land Mission'.lua")
```
Comment (tridge) [inline libraries/AP_Scripting/applets/UniversalAutoLand.lua:122]: use MAV_SEVERITY table, as per other scripts, also typo with the ' ?

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_Scripting/applets/UniversalAutoLand.lua (hunk 3 above)
Comment (tridge) [reply libraries/AP_Scripting/applets/UniversalAutoLand.lua:122]: maybe should say UniversalAutoland ?

File: libraries/AP_Scripting/applets/UniversalAutoLand.lua [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+  if not arming:is_armed() then --if disarmed, wait until armed
+    mission_loaded = false
+    return update,1000
+  end
+  
+  if not mission_loaded then --if first time after arm and enabled is valid then create  mission 
+    local home = ahrs:get_home()
+    local location = ahrs:get_location()
+    if location and home and location:alt() > (home:alt() + param:get("TKOFF_LVL_ALT")*100) then
+        local yaw = ahrs:get_yaw()
```
Comment (tridge) [inline libraries/AP_Scripting/applets/UniversalAutoLand.lua:113]: likely want ground course

Comment (tridge) [reply libraries/AP_Scripting/applets/UniversalAutoLand.lua:113]: actually, should be based on launch pt and pt when it has climbed up

Comment (Hwurzburg) [reply libraries/AP_Scripting/applets/UniversalAutoLand.lua:113]: I disagree...I want a 180 reciprocal waypoint from the initial takeoff heading, which in 99% of the cases should be into the wind or down the runway...so not ground course which could be heavily skewed by crosswinds...

File: libraries/AP_Scripting/applets/UniversalAutoLand.lua (hunk 3 above)
Comment (Hwurzburg) [reply libraries/AP_Scripting/applets/UniversalAutoLand.lua:122]: will change

File: libraries/AP_Scripting/applets/UniversalAutoLand.lua (hunk 4 above)
Comment (Hwurzburg) [reply libraries/AP_Scripting/applets/UniversalAutoLand.lua:113]: ground course on takeoffs with crosswind will set the final approach askew to the runway....

File: libraries/AP_Scripting/applets/UniversalAutoLand.lua (hunk 2 above)
Comment (Hwurzburg) [reply libraries/AP_Scripting/applets/UniversalAutoLand.lua]: done

File: libraries/AP_Scripting/applets/UniversalAutoLand.lua (hunk 3 above)
Comment (Hwurzburg) [reply libraries/AP_Scripting/applets/UniversalAutoLand.lua]: done

File: libraries/AP_Scripting/applets/UniversalAutoLand.lua (hunk 4 above)
Comment (Hwurzburg) [reply libraries/AP_Scripting/applets/UniversalAutoLand.lua:148]: my point:
image

Comment (Hwurzburg) [reply libraries/AP_Scripting/applets/UniversalAutoLand.lua]: @tridge After our conversation, I have switched to setting bearing quickly using GPS ground course at .5 AIRSPEED_MIN and it seems to work well, acquiring bearing before rolling lift off and well before TKOFF_LVL_ALT, so should suffice for both compass and compass-less takeoffs...will flight test this week

**Comment by Hwurzburg** [comment]
I test flew on compass-less plane with strong wind...both into the wind and at 90deg crosswind....in both cases the created final approach waypoint was place in line with throw direction....much better crosswind performance than waiting for LVL_ALT, since by then the yaw has weathervaned into the wind rather than being the initial throw/roll-out direction...

https://youtu.be/nd9rq2448Hw

**Review by tridge — CHANGES_REQUESTED** [review]
I think an autotest that checks we end up close to home and pointing in the reverse direction (within a margin) of the takeoff direction would be worthwhile

**Comment by Hwurzburg** [comment]
An autotest is gating a Lua applet?

**Review by tridge — APPROVED** [review]

File: libraries/AP_Scripting/applets/UniversalAutoLand.lua [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+assert(param:add_table(PARAM_TABLE_KEY, PARAM_TABLE_PREFIX, 4), 'could not add param table')
+
+-- add a parameter and bind it to a variable
+function bind_add_param(name, idx, default_value)
+    assert(param:add_param(PARAM_TABLE_KEY, idx, name, default_value), string.format('could not add param %s', name))
+    return Parameter(PARAM_TABLE_PREFIX .. name)
+end
+
+--[[
+  // @Param: ENABLE
```
Comment (tridge) [inline libraries/AP_Scripting/applets/UniversalAutoLand.lua:16]: need full name of params in lua scripts, or you end up with just "ENABLE" as the param

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_Scripting/applets/UniversalAutoLand.lua (hunk 5 above)
Comment (Hwurzburg) [reply libraries/AP_Scripting/applets/UniversalAutoLand.lua]: done

**Review by tridge — APPROVED** [review]

**Pull request merged**